### ISSUE TITLE
Alt keys losing focus when another key is pressed

### ISSUE THREAD

**Issue by jayvogt** [rvaiya/keyd#207]
Maybe I'm just going crazy, but I can't for the life of me figure out what's going on. I'm just trying to switch my leftalt and leftmeta keys. With my current configuration, leftshift, leftcontrol, rightshift, and rightcontrol all work correctly if I hold and press another key (like leftcontrol + c). If I try to open the Terminal with my shortcut leftcontrol + leftalt + "T", leftalt clears as soon as I hit "T" (or any other key for that matter...)

```
[ids]

*

[main]

leftalt = leftmeta
leftmeta = leftalt

# Map right-side keys to themselves, otherwise they act as their left counterparts
rightalt = rightalt
rightcontrol = rightcontrol
rightshift = rightshift
```

Am I doing something wrong?

**Comment by rvaiya** [comment]
Yes, mapping directly to modifier keys interferes with keyd's modifier layer logic. If you have inspected the logs you would have seen a warning about this, though it is probably my fault for not documenting it better. 

You probably want:

```
alt = layer(meta)
meta = layer(alt)
```

unless you are doing something unorthodox. Mapping the right modifiers to themselves is probably a bad idea unless you have a good reason for doing this. Most applications use modifier masks which don't distinguish between their left and right activation keys. See also #85 and #86.

**Comment by jayvogt** [comment]
Wow. I'm embarrassed now... I saw the logs a lot while trying to troubleshoot, but I just realized I didn't notice that systemctl status was cutting off the "you probably want layer(<modifier name>) instead."

Also, I'm trying to still wrap my head around the layer functionality. Am I correct in understanding that the layer function is a non-destructive way of modifying keys by adding instructions to cover up the native key function, whereas I was attempting to replace the native key function instead?

Also, thanks for the references about the right-side keys. I had no specific need for them to be separate, I just assumed it would be best if they all functioned as their particular keys. I'll leave them alone unless I run into a reason to map them.

Thank you so much for your work on this project and your help. I really appreciate this utility!

**Comment by rvaiya** [comment]
> Am I correct in understanding that the layer function is a non-destructive way of modifying keys by adding instructions to cover up the native key function, whereas I was attempting to replace the native key function instead?

Layers are just transiently active keymaps. If you are familiar with the concept of a layer in other remapping tools (e.g QMK) the concept is the same in keyd. 

The main different between keyd and other remapping tools is that keyd has a notion of 'modifier layers'  which can be conceived of as layers which inherit the properties of the modifiers from which they are descended.

For example, you might define a custom control layer like so:

```
[main]
capslock = layer(mycontrol)

[mycontrol:C]
a = A-tab
``` 

which will cause `capslock` to behave like control except when `a` is pressed, in which case it will emit `A-tab`. By default modifier keys (e.g `leftshift`) are mapped to empty modifier layers of the same name (e.g `shift`) which allows you to do things like this:

```
[shift]
c = C-c
```

without having to explicitly remap `leftshift`.

The upshot of all this is that you can (and should) treat modifiers as layers within keyd (with some caveats).

**Comment by jayvogt** [comment]
Ah, okay. That makes a lot more sense. Thank you for taking the time to explain this to me!
